### PR TITLE
Add KDD for versions and compatibility

### PR DESCRIPTION
**Pull request by andreievg**
Would require a placeholder issue for implementation of this before merging

### CODE REVIEW

File: decisions/version-compatibility.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+* 1.07 remote is compatible with `>=` 5.08 central (this is carried over from 1.02 verision of remote).
+* 5.08 is compatible with `>=` 1.02 of remote
+* 6 central is compatible with `>=` 1.02 of remote  (this is carried over from 5.07 verision of central).
+* 2 of remote is compatible with `>=` 6 of central
+
+So let's say 1.07 is trying to sync with 5.08, central says it's compatible with 1.02 and up and 1.07 knows it's compatible with 5.08 and up, thus it's ok to sync.
+
+2 is trying to sync with 5.08, central says it's compatible with 1.02 and up and remote knows it's compatible with 6 and up, sync is not allowed.
+
+1.02 is tyring to sync with 5.08. central says it's compatible with 1.02 and up and remote knows it's compatible with 5.06 and up, sync is not allowed
```
Comment (mark-prins) [inline decisions/version-compatibility.md:117]: 🤔  in this example I think sync would be allowed. remote is compatible with 5.06 and up; and the central version is 5.08 which is 'up' from 5.06

File: decisions/version-compatibility.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+*Pros:*
+- Simpler then `Option 2` and slighly more invovled then `Option 1`, but seems (subjectively) the best balance for compatibility/complexity
+- Quite flexible in forward and backwards compatibility
+
+*Cons:*
+- More involved, and flexibility can be a doulbe edge sword
```
Comment (mark-prins) [inline decisions/version-compatibility.md:124]: soz.
:nit
```suggestion
- More involved, and flexibility can be a double edged sword
```

**Review by mark-prins — APPROVED** [review]
🎉  awesome - thanks. 
Nothing much to add.. I think that we need to have the compatibility in both sides of the API contract to account for upgraded versions trying to communicate with older ones. Otherwise, like you say, it's too rigid and the allowed versions is very limited.

One concern is the maintenance of the compatibility info - but that's a separate problem. We require devs to update when there's a breaking change and we can't always rely on devs to remember these things 🙄

File: decisions/version-compatibility.md (hunk 2 above)
Comment (andreievg) [reply decisions/version-compatibility.md:124]: 🙏

File: decisions/version-compatibility.md (hunk 1 above)
Comment (andreievg) [reply decisions/version-compatibility.md]: Oops, 🙏 for the catch
I think I was trying to show an example where remote thinks it's compatible but central disagrees, changed the example: 8cb9f5d31f7c318168ab26ccd639129ca4ee6343

and added headers: fa9904203bea2d7c604b7d7aaa562217b5e3fb21

**Comment by andreievg** [comment]
Thanks @mark-prins

**Comment by Chris-Petty** [comment]
Sorry didn't see till now. I'd differentiate these as 2 different breaking changes:

> ###  Examples of breaking compatibility changes
> 1. New sync field is added to a table on remote server, it doesn't exist on central server, causing errors during sync `push` (1.02 <-> 5.06 example in table below)
> 2. Sync api shape has changed on central server causing errors during `pull` (1.01 <-> 5.08 example in table below)

First this is a _DB schema_ breaking change. Effectively, this is managing the JSON schema changes within each "sync record". Central schema should be incremented, say up to v36. Old remotes should say "I'm expecting DB schema v35". Central checks incoming requests, if it's known that v35 is going to break with v36, the central server can reject. Likewise if the remote is ahead of the central server, the remote should be rejected by the central server until the central server is upgraded (it'd be rare for central not be the first site upgraded anyway). Generally, adding fields and tables on the central server will be backwards compatible, modifying or removing fields will not. If a remote site is missing fields that are on the central server, it risks data being lost from the system (e.g. if you didn't have price fields on a invoice_line record. You might think that's a crazy example, but that's what happened with mobile for literally years! When pushed back to the server the field data is lost for good). If the remote has any fields/tables that are not yet on the central server, its data isn't being fully preserved.

Second is an API breaking change, where we need to manage the shape of the sync record itself, headers, authorisation or the body shape of the response, or error shapes. If we ever make a backwards incompatible change to the sync bodies accepted and returned, we should version the API and do it though a new route to ensure older remotes can keep syncing (e.g. mSupply desktop old sites on sync v1, mobile on sync v2 and v3).

Pros:
- I think this gives clearer feedback to the remote site developers what is wrong 
- The decision to accept a request is more firmly on the central server
- Allows for maintaining stable APIs (particularly with automated testing), ensuring backwards compatibility until we intentionally and clearly deprecate and eventually remove an older API (e.g. eventually sync v1).
- If we were to lean heavily into stable APIs into the future, the structure does allow for implementing mapping functions
- Keeps us from adding app versions into the mix of API versioning, when we already have API versioning. This is particularly important with an open API where anyone could be writing a client - we don't want to have to maintain every possible open source client's version in the central server code base.

Cons:
- We have to be intentional about changes that affect interoperability (I think this is a pro)
- Introduces schema versioning to the databases, something we'll have to do. (Similar to this, Realm DB does this check between schema changes on app version upgrades to ensure migrations are done both internally by the DB and to give an opportunity for programatic changes)

**Comment by andreievg** [comment]
Thanks Chris, I have a few questions:

> Introduces schema versioning to the databases, something we'll have to do

Do we ever not increment mSupply versions when schema is changed ?

mSupply version is already coupled to schema version afaik.

For api versions, if we modify V5 slightly (like aligning central and remote record shape), would we need to increment V5 or add V6 ? 

Have you considered `complexity` as a con ? (having to consider multiple types of versions, and making sure they are always incremented).

I think the suggestion of seperate API compatibility and schema compatibility is quite good, the schema compatibility would still need to be recorded on both remote and central, since even though this is true -> `it'd be rare for central not be the first site upgraded anyway` we have a lot less control of remote/mobile instances (they can be upgraded without central).

I don't quite understand what you proposal is with API version, can you give me a use case of how it works (is it in headers ? is it in the route ?)

**Comment by Chris-Petty** [comment]
> Thanks Chris, I have a few questions:
> 
> > Introduces schema versioning to the databases, something we'll have to do
> 
> Do we ever not increment mSupply versions when schema is changed ?
> 
> mSupply version is already coupled to schema version afaik.
> 

We do increment with schema changes, but also if we just change the width of a form in the UI. There is ambiguity in that mSupply version does not clearly indicate whether there has been a schema change _or not_ so we lose feedback for API consumers.

> For api versions, if we modify V5 slightly (like aligning central and remote record shape), would we need to increment V5 or add V6 ?
> 

Yes this is typical of API versioning if you make breaking changes. I am hesitant about API shape changes without keeping previous versions accessible and stable with automated tests, or closing down and rejecting. In my experience it is very time consuming to test even just _recently_ past versions of mSupply Desktop and mSupply Mobile for compatibility with every new release of the mSupply server, not to mention past mSupply server versions. I believe this is required with option 3 and it'll be a massive time and capacity requirement needed to maintain an accurate central server lookup of compatibility of all versions of all client applications. Automated integration tests from all potential clients is complex and time consuming to implement and maintain (while this arguably should be done regardless, we can't guarantee code ownership of all clients...).

To put my proposal another way, a tweak you could make to option 3 (or option 2) is for the remote to specify what version of the _central server application_ it's expecting, rather than its own version. The central server would check this and decide whether the expected version is compatible with its current version. If we had a fairly comprehensive automated test suite just on the central server, every time it breaks we fix the test and start rejecting all previous versions of the server that is being requested. I think this has a large benefit over the current solution, in that the server codebase does not need to know about client code bases and versioning. 

Sending the expected DB schema version has similar merit, is largely the same as sending the expected server version but more clearly communicated what the point is. 

> Have you considered `complexity` as a con ? (having to consider multiple types of versions, and making sure they are always incremented).
> 

It is a con, but I think the development complexity and maintenance is less than option 3 as we're testing less combinations, over less boundaries, and every remote app version does not require updating the server codebase. I think it is more typical API design as well, which has value.

> I think the suggestion of seperate API compatibility and schema compatibility is quite good, the schema compatibility would still need to be recorded on both remote and central, since even though this is true -> `it'd be rare for central not be the first site upgraded anyway` we have a lot less control of remote/mobile instances (they can be upgraded without central).
> 

I agree that remotes should record and send what schema it's expecting, but not with the server tracking schemas of remotes. (At least not for logic, informationally is great.)

> I don't quite understand what you proposal is with API version, can you give me a use case of how it works (is it in headers ? is it in the route ?)

Sync API: `sync/v5/`, new versions would increment the API there with new routes. Old version routes would stay open until we don't want to maintain them anymore. Old apps using closed API versions are rejected. New apps using  API versions that aren't yet available are rejected.

For a DB schema it'd be in the headers or a handshake. If the DB version is different either:
- The client expects a newer version of the DB schema than the server uses: reject the client until the server upgrades (I expect this to be an edge case)
- The client expects an older version of the DB schema. The new schema belongs to a newer version of the server. Automated tests in development of that newer server should detect incompatible schemas, probably by posting older shapes against the new shapes. We choose either mapping the data to be compatible with the latest versions or rejecting older versions. For instance, a field/table name change can trivially be mapped from old to new. A refactoring of an object field to some tables could be mapped, or we just say the remote needs to upgrade and migrate itself before it continues syncing.
  - You could do the same with app version but it'd be less clear what's going on.
